istresearch/traptor
# P4 2412 - Tweet Sampling

# For the Reviewer

When this is complete, you should approve the PR via github.

# For the Reviewee

<Reminder PR Title should be JIRA_NUMBER and Useful Description>

## Summary
Description of changes.

#### Release Note
<Concise sentence describing change>Required.

#### Breaking Changes
<Description for Techops of how to handle changes, migrations, updates>None.

## Quality Assurance

You have gathered the following items:

Given all of the items above, you have updated your Application ORT at the following locations:
- **Features and Alerting**: <link to app ORT>Required.
- **P4 Alerting**: <link to p4-alerting PR>Required.

## Testing and Verification

*Steps to test your application for someone not familiar with it.* Required.

1. Start the supporting stack:
```
$ docker-compose up -d kibana logstash redis kafka elasticsearch
```
2. Build and start traptor:
```
docker-compose up -d --build traptor
```
3. In a separate terminal, stream the kafka topic:
```
$ docker-compose exec kafka /opt/kafka/bin/kafka-console-consumer.sh --bootstrap-server kafka --zookeeper zookeeper --topic traptor
```
4. Visit http://localhost:5601/app/discover#/ and create an index pattern for `traptor*` using the `timestamp` field. Then go to Discover and ensure that Traptor is waiting for rules.
5. Task Traptor by running the `add_rules.py` script:
```
# pip install redis
$ python ./scripts/add_rules.py
```
5. Then go back Discover and verify that Traptor begins collecting data. It may take up to 10 seconds to detect the rule tasking.
6. Now query for this message: `message.keyword : ("Twitter Rate" or "Kafka Rate" )`. Add columns for `rule_value`, `average_tps`, and `count`. It should be apparent that the volume of Tweets received from Twitter is higher than thee volume of tweets Traptor is writing to Kafka.
7. Confirm tweets are written to the Kafka topic by watching the terminal established in step 3.

8. Run Traptor in each of it's three modes by editing the `TRAPTOR_TYPE` env variable in the compose file. The three modes are `track`, `follow`, and `locations`.
```
# Edit mode in compose, then run this...
$ docker-compose up -d traptor
```

If you build a viz, it could look like this:
![Screen Shot 2021-03-24 at 11 11 34 AM](https://user-images.githubusercontent.com/7338230/112336504-82f35b80-8c93-11eb-8be1-e5d60f3d90da.png)
